### PR TITLE
Add Back to top block

### LINKED ISSUE #50433: Create a "Back to Top" Block
## What problem does this address?

The popularity of responsive web design has led to a proliferation of single-column, long-page designs both on mobile and desktop.  A consequence of these designs has been the **Back to Top** button, which is a shortcut that allows users to quickly navigate to the top of the page.

## What is your proposed solution?

I suggest adding this feature to the site editor, perhaps it makes sense for it to be a block, instead of a hidden feature in the Button block.

I expect it to be fully customisable (colors, typography, background, sticky vs non-sticky, etc) here are a few example designs:

Captura de ecrã 2023-05-08, às 14 52 26

Captura de ecrã 2023-05-08, às 14 55 55

Captura de ecrã 2023-05-08, às 15 06 30

cc: @iamarinoh @iamtakashi @jasmussen

### PR DESCRIPTION
**Pull request by carolinan**
## What?

Updated October 11.

Adds a basic block with a paragraph containing a link that takes you to the top of the browser window.
Fixes https://github.com/WordPress/gutenberg/issues/50433

## Why?

From the issue:
"The popularity of responsive web design has led to a proliferation of single-column, long-page designs both on mobile and desktop. A consequence of these designs has been the Back to Top button, which is a shortcut that allows users to quickly navigate to the top of the page."

## How?

Adds a new block called Back to top.
- The block has a single block attribute which is the link text.
- It uses RichText with the link feature disabled.
- It sets HTML editing to false in block.json.
- The link in the editor is not functional, and that is intended. Selecting it is only for editing the text.
- The target id, `wp-block-back-to-top`, is added to front of the website so that the block always have a target on the page.
- The JavaScript file is needed to move focus, so that the next tab stop is the first focusable element of the page.

## Considerations

- I have placed this block in the "design" category. This is probably wrong. Perhaps there should be a "utility" category?
- I did not disable the inline image on the RichText. I am concerned that the user will only add an image and no text to use as the link text.

## Testing Instructions

1. Add a back to top block, for example in the site footer.
3. Add a second back to top block: change the text, make it part of it bold, italic etc, add an inline image.
4. Save and confirm that both links display correctly on the front.
5. When activating the link, the view should scroll to the top of the page. 

## Updated Testing Instructions for keyboard
1. To add the block, type in, for example, "back" or "top," and the block can be selected from the shortcut.
2. When the block is added, the focus is at the start of the editable text field. The text can be edited, and the default text can be removed.
3. Pressing shift + tab will move the focus to the block toolbar.
4. Navigate the toolbar with the arrow keys to transform and move the block, style selected text, and open the options menu.
5. On the front, navigate to the link and activate it.
6. The view should scroll to the top of the page.
7. Pressing the tab key should move focus to the first focusable element in the body, though after the WordPress navigation bar.

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/7422055/149d49c2-87b3-49b6-b0c7-a9ca2a6a7bb5

### CODE REVIEW

**Comment by alexstine** [comment]
@carolinan Is this only a scroll function being used here? If so, you need to find a way to manage focus after the scroll transition for screen readers or the link will be entirely useless for accessibility. Something like this?

```
document.querySelector('body a').focus();
```

That would place focus on the first link after the opening `<body>` tag. The problem with a block like this is its great for the sighted but it will be a real challenge trying to figure out how to give an equal experience for everyone since a reliable anchor will not be possible.

Thanks.

**Comment by carolinan** [comment]
Maybe an ID could be added to the automated skip to content link, and that could be the target.

**Comment by alexstine** [comment]
@carolinan Sounds good. In that case, we'd have to restrict this block to block themes only I guess so that way we know for sure the ID will be on the skip link.

**Comment by carolinan** [comment]
Sigh, there is always something 😆

**Comment by alexstine** [comment]
> Sigh, there is always something 😆

I hear that, hashtag LifeOfDevs.

**Comment by carolinan** [comment]
I have added a basic viewScript. It doesn't contain any "scroll to" functionality, it only moves focus when clicking on the back to top block.

**Comment by hanneslsm** [comment]
should smooth control be included directly as an option?
https://developer.mozilla.org/en-US/docs/Web/CSS/scroll-behavior

Not sure how easy it is to implement, though.

File: packages/block-library/src/back-to-top/edit.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { useBlockProps, RichText } from '@wordpress/block-editor';
+
+export default function BackToTopEdit( { attributes, setAttributes } ) {
+	const { text } = attributes;
+	return (
+		<p { ...useBlockProps() }>
+			<RichText
```
Comment (alexstine) [inline packages/block-library/src/back-to-top/edit.js:11]: This needs some type of label to explain to users that this will be the text of the link.

**Review by alexstine — COMMENTED** [review]
@carolinan This is working well. Just one note below.

File: packages/block-library/src/back-to-top/edit.js (hunk 1 above)
Comment (carolinan) [reply packages/block-library/src/back-to-top/edit.js:11]: Added, thank you.

**Comment by aristath** [comment]
What if, instead of adding a viewScript, the link had `onClick="document.querySelector( '.skip-link' ).click()"` ? 🤔

**Comment by carolinan** [comment]
Thank you for the suggestion. 
The example only moves focus to the correct focusable element if the target exists on the page. 
To move to either the first focusable element after the target id, or the first focusable element in the body, there would need to be a condition, and I am not sure if something like this would be an improvement? It is longer than I would like for adding JS inline.
`'<p %1$s><a href="#top" onClick="if( document.getElementById( \'top\' ) ) { document.getElementById( \'top\' ).click() } else { document.querySelector( \'body a\' ).focus();}">%2$s</a></p>',
`
Users cannot change the target id, but I have no doubt that some will add that anchor (top), to the element they want the link to skip to and that might not be the top of the page, it can be the top of an article or similiar.

**Comment by carolinan** [comment]
During a coding session with @SergeyBiryukov, @aristath, and @afercia, we discussed trying a solution with no JavaScript on the front, by adding the target id to the top of the page.

**Comment by alexstine** [comment]
If we're moving away from the view.js file, does this mean this block will only be available for a block theme? That is the only possible way we'd have any idea at all if that ID exists. It would still be a gamble at best.

**Comment by carolinan** [comment]
> If we're moving away from the view.js file, does this mean this block will only be available for a block theme? That is the only possible way we'd have any idea at all if that ID exists. It would still be a gamble at best.

The id is added to the front if the theme uses [wp_body_open()](https://developer.wordpress.org/reference/functions/wp_body_open/), regardless of theme type.

**Comment by alexstine** [comment]
Do we have any data to support saying most 3rd-party themes use this hook? Any data that says all default themes use this hook?

**Comment by carolinan** [comment]
I don't have access to third party stats. 
All classic themes submitted to the WordPress.org theme directory from june 2019 are required to include it. 
I can check if Themeforest requires it.

**Comment by carolinan** [comment]
Themeforest does not require it.
[Reference on envato.com.](https://help.author.envato.com/hc/en-us/articles/360000479946-WordPress-Theme-Requirements-Part-4-Coding#h_3677526642841522300409445)

**Comment by alexstine** [comment]
I think we should keep the view.js script as a fallback then. It is a couple lines of JS, should not be a deal breaker.

**Comment by carolinan** [comment]
Setting as draft again while I explore if it can and should use the interactivity API.

**Comment by carolinan** [comment]
@beafialho @WordPress/gutenberg-design @mtias It has been a couple of months without reviews, change requests or comments, is this still a block that is on the roadmap, that we want to include?
I am pushing because I would like to free up my time to work on other issues.

**Comment by beafialho** [comment]
This looks good to me. Adding a bit of (likely late) feedback: perhaps in an upcoming iteration the block can eventually evolve into “button” territory to surface some of the design references described in [this issue](https://github.com/WordPress/gutenberg/issues/50433). It seems like making it a button is a little less about the semantics, and more about the block supports (position, background, padding, radius, text color, etc.). Even if there's no possibility to add an icon soon, it could be useful to be able to customize the text "Back to top", make it sticky, etc.

**Comment by jasmussen** [comment]
Agreed with Bea, if the current iteration does not preclude us from future customization supports, I could see a first version landing of this. Thanks for the work!

**Comment by jameskoster** [comment]
Forgive my naivety but why does this need to be a dedicated block? It's reasonably straight-forward to link any text (or button) to `#top` already using the link control UI.

**Comment by paaljoachim** [comment]
Could we instead just add an option to the Buttons block?
As to me it seems like a link that could be inside a button or just a link. (A button could just show a regular text link.)
It seems kind of weird to have this as its own block.

**Comment by carolinan** [comment]
Closing this as not planned.

**Comment by paaljoachim** [comment]
Hi

I have been looking at this feature again.... I added my above comment thinking that we could add simple ways to a link or Button block but I am starting to change my mind.
I also came across this back to top block plugin and also tried it out. I went ahead and added it to the Post template.
https://wordpress.org/plugins/makeiteasy-back-to-top/

I am wondering if including a native Gutenberg block could be a good idea. As it keeps it simple and straight forward and it is a feature used a lot in posts that have long layouts. Over time it could also be adjusted based on additional needs for this specific block. Should we reopen?

**Pull request closed without merging**